File: solutions/erlang/collatz-conjecture/1/src/collatz_conjecture.erl
% For my Junko F. Didi and Shree DR.MDD

-module(collatz_conjecture).

-export([steps/1]).

-spec steps(integer()) -> non_neg_integer() | error.
steps(Value) ->
    case is_integer(Value) andalso Value > 0 of
        true -> steps(Value, 0);
        _ -> error(badarg)
    end.

-spec steps(integer(), non_neg_integer()) -> non_neg_integer().
steps(Value, Count) ->
    case Value of
        1 -> Count;
        _ -> if
                Value band 1 =:= 0 -> steps(Value bsr 1, Count + 1);
                true -> steps(3 * Value + 1, Count + 1)
            end
    end.

File: solutions/fortran/collatz-conjecture/1/collatz_conjecture.f90
module collatz_conjecture
   implicit none
contains
   integer function steps(i)
      integer :: i
      integer :: current
      if (i <= 0) then
         steps = -1
         return
      end if
      current = i
      steps = 0
      do while (current /= 1)
         steps = steps + 1
         if (mod(current, 2) == 0) then
            current = current/2
         else
            current = 3*current + 1
         end if
      end do
   end function
end module